pkg/apis/admission/types.go

```diff
@@ -19,58 +19,73 @@ package admission
 )
 
 // +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
 
-// AdmissionReview describes an admission request.
+// AdmissionReview describes an admission review request/response.
 type AdmissionReview struct {
 	metav1.TypeMeta
 
-	// Spec describes the attributes for the admission request.
-	// Since this admission controller is non-mutating the webhook should avoid setting this in its response to avoid the
-	// cost of deserializing it.
-	Spec AdmissionReviewSpec
-	// Status is filled in by the webhook and indicates whether the admission request should be permitted.
-	Status AdmissionReviewStatus
+	// Request describes the attributes for the admission request.
+	Request AdmissionRequest
+
+	// Response describes the attributes for the admission response.
+	// +optional
+	Response AdmissionResponse
 }
 
-// AdmissionReviewSpec describes the admission.Attributes for the admission request.
-type AdmissionReviewSpec struct {
+// AdmissionRequest describes the admission.Attributes for the admission request.
+type AdmissionRequest struct {
+	// UID is an identifier for the individual request. It allows us to distinguish instances of requests which are
+	// otherwise identical (parallel requests, requests when earlier requests did not modify etc)
+	UID types.UID
 	// Kind is the type of object being manipulated.  For example: Pod
 	Kind metav1.GroupVersionKind
-	// Name is the name of the object as presented in the request.  On a CREATE operation, the client may omit name and
-	// rely on the server to generate the name.  If that is the case, this method will return the empty string.
-	Name string
-	// Namespace is the namespace associated with the request (if any).
-	Namespace string
-	// Object is the object from the incoming request prior to default values being applied
-	Object runtime.Object
-	// OldObject is the existing object. Only populated for UPDATE requests.
-	OldObject runtime.Object
-	// Operation is the operation being performed
-	Operation Operation
 	// Resource is the name of the resource being requested.  This is not the kind.  For example: pods
 	Resource metav1.GroupVersionResource
 	// SubResource is the name of the subresource being requested.  This is a different resource, scoped to the parent
 	// resource, but it may have a different kind. For instance, /pods has the resource "pods" and the kind "Pod", while
 	// /pods/foo/status has the resource "pods", the sub resource "status", and the kind "Pod" (because status operates on
 	// pods). The binding resource for a pod though may be /pods/foo/binding, which has resource "pods", subresource
 	// "binding", and kind "Binding".
+	// +optional
 	SubResource string
+	// Name is the name of the object as presented in the request.  On a CREATE operation, the client may omit name and
+	// rely on the server to generate the name.  If that is the case, this method will return the empty string.
+	// +optional
+	Name string
```

Comment:
This is violating our API convention that the empty value of an optional field has to be `nil`. But I guess we don't care anymore since there are so many violations in our code base.

Reply:
This is reflecting the optionality of what was already in the API. I agree we need to come up with a consistent story on what and when we are going to do about our large base of optional/non-pointer fields.

Reply:
"If that is the case, this method will return the empty string." ...so, it's not really optional, is it?